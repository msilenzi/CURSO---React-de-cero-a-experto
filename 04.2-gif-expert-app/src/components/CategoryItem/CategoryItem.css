.masonry-container {
  padding: 0;
  column-count: 3;
  column-gap: 1rem;
  width: 100%;
}

@media (max-width: 900px) {
  .masonry-container {
    column-count: 2;
  }
}

@media (max-width: 560px) {
  .masonry-container {
    column-count: 1;
  }
}

.masonry-item {
  padding: 0;
  list-style: none;
  border-radius: 0.75rem;
  margin-bottom: 1rem;
  background-color: #fff;
  padding: 0.5rem;
  break-inside: avoid;
  box-shadow: 0 0 1rem rgba(0, 0, 0, 0.15);
}

.masonry-item__image {
  object-fit: contain;
  border-radius: 0.5rem;
  width: 100%;
}

.masonry-item__details {
  margin-top: 0.5rem;
}

.masonry-item__details--user {
  display: grid;
  grid-template-columns: 40px auto;
  column-gap: 0.5rem;
}

.masonry-item__avatar {
  grid-row: 1 / span 2;
  border-radius: 100%;
}

.masonry-item__title {
  margin: 0;
  font-weight: 700;
  color: var(--primary-900);
}

.masonry-item__username {
  margin: 0;
  color: var(--primary-100);
}
