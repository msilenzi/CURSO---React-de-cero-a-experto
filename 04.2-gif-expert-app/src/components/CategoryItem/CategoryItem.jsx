import PlaceholderedImage from '../../ui/PlaceholderedImage'
import './CategoryItem.css'

function CategoryItem({ title, image, user }) {
  return (
    <li className="masonry-item">
      <ItemImage title={title} image={image} />
      <ItemDetails title={title} image={image} user={user} />
    </li>
  )
}

function ItemImage({ title, image }) {
  const rndColor = Math.floor(Math.random() * 5) + 2

  return (
    <a
      className="masonry-item__image"
      href={image.url}
      target="_blank"
      rel="noopener noreferrer"
    >
      <PlaceholderedImage
        className="masonry-item__image"
        placeholderColor={`var(--primary-${rndColor}00)`}
        src={image.url}
        alt={title}
        height={image.height}
        width={image.width}
      />
    </a>
  )
}

function ItemDetails({ title, image, user }) {
  if (!user) {
    return (
      <div className="masonry-item__details masonry-item__details--no-user">
        <ItemTitle title={title} image={image} />
      </div>
    )
  }

  return (
    <div className="masonry-item__details masonry-item__details--user">
      <ItemAvatar user={user} />
      <ItemTitle title={title} image={image} />
      <ItemUsername user={user} />
    </div>
  )
}

function ItemTitle({ title, image }) {
  return (
    <a
      className="masonry-item__title"
      href={image.url}
      target="_blank"
      rel="noopener noreferrer"
    >
      {title}
    </a>
  )
}

function ItemAvatar({ user }) {
  return (
    <a
      className="masonry-item__avatar"
      href={user.profile_url}
      target="_blank"
      rel="noopener noreferrer"
    >
      <PlaceholderedImage
        placeholderColor={'var(--primary-100)'}
        src={user.avatar_url}
        alt={`${user.username} avatar`}
      />
    </a>
  )
}

function ItemUsername({ user }) {
  return (
    <a
      className="masonry-item__username"
      href={user.profile_url}
      target="_blank"
      rel="noopener noreferrer"
    >
      {user.username}
    </a>
  )
}

export default CategoryItem
