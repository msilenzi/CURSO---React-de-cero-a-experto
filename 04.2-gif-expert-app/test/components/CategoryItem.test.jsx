import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import CategoryItem from '../../src/components/CategoryItem/CategoryItem'

describe('Pruebas en <CategoryItem />', () => {
  const mockTitle = 'Pokemon'
  const mockImage = {
    url: 'https://giphy.com/images/pokeom.gif',
    height: 800,
    width: 600,
  }
  const mockUser = {
    profile_url: 'https://giphy.com/profiles/pokemon',
    avatar_url: 'https://giphy.com/avatar/pokemon',
    username: 'Pokemon Company',
  }

  test('debe coincidir con el snapshot: con usuario', () => {
    const { container } = render(
      <CategoryItem title={mockTitle} image={mockImage} user={mockUser} />
    )
    expect(container).toMatchSnapshot()
  })

  test('debe coincidir con el snapshot: sin usuario', () => {
    const { container } = render(
      <CategoryItem title={mockTitle} image={mockImage} />
    )
    expect(container).toMatchSnapshot()
  })

  test('debe mostrar el título untitled', () => {
    render(<CategoryItem title="" image={mockImage} user={mockUser} />)
    screen.getByText('Untitled')
  })

  test('debe mostrar la información del usuario', () => {
    render(<CategoryItem title={mockTitle} image={mockImage} user={mockUser} />)
    screen.getByText(mockUser.username)
  })

  test('debe mostrar la información imagen correctamente', () => {
    render(<CategoryItem title={mockTitle} image={mockImage} />)
    const image = screen.getByAltText(mockTitle)
    expect(image).toHaveAttribute('src', mockImage.url)
    expect(image).toHaveAttribute('height', mockImage.height.toString())
    expect(image).toHaveAttribute('width', mockImage.width.toString())
  })
})
