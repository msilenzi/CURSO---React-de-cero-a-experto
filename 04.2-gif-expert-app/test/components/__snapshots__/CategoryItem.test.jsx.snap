// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <CategoryItem /> debe coincidir con el snapshot: con usuario 1`] = `
<div>
  <li
    class="masonry-item"
  >
    <a
      class="masonry-item__image"
      href="https://giphy.com/images/pokeom.gif"
      rel="noopener noreferrer"
      target="_blank"
    >
      <div
        class="placeholdered-image"
        data-testid="PlaceholderedImage"
      >
        <img
          alt="Pokemon"
          class="masonry-item__image"
          height="800"
          src="https://giphy.com/images/pokeom.gif"
          style="visibility: hidden;"
          width="600"
        />
      </div>
    </a>
    <div
      class="masonry-item__details masonry-item__details--user"
    >
      <a
        class="masonry-item__avatar"
        href="https://giphy.com/profiles/pokemon"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          class="placeholdered-image"
          data-testid="PlaceholderedImage"
        >
          <img
            alt="Pokemon Company avatar"
            src="https://giphy.com/avatar/pokemon"
            style="visibility: hidden;"
          />
        </div>
      </a>
      <div
        class="masonry-item__user"
      >
        <a
          class="masonry-item__title"
          href="https://giphy.com/images/pokeom.gif"
          rel="noopener noreferrer"
          target="_blank"
        >
          Pokemon
        </a>
        <a
          class="masonry-item__username"
          href="https://giphy.com/profiles/pokemon"
          rel="noopener noreferrer"
          target="_blank"
        >
          Pokemon Company
        </a>
      </div>
    </div>
  </li>
</div>
`;

exports[`Pruebas en <CategoryItem /> debe coincidir con el snapshot: sin usuario 1`] = `
<div>
  <li
    class="masonry-item"
  >
    <a
      class="masonry-item__image"
      href="https://giphy.com/images/pokeom.gif"
      rel="noopener noreferrer"
      target="_blank"
    >
      <div
        class="placeholdered-image"
        data-testid="PlaceholderedImage"
      >
        <img
          alt="Pokemon"
          class="masonry-item__image"
          height="800"
          src="https://giphy.com/images/pokeom.gif"
          style="visibility: hidden;"
          width="600"
        />
      </div>
    </a>
    <div
      class="masonry-item__details masonry-item__details--no-user"
    >
      <a
        class="masonry-item__title"
        href="https://giphy.com/images/pokeom.gif"
        rel="noopener noreferrer"
        target="_blank"
      >
        Pokemon
      </a>
    </div>
  </li>
</div>
`;
